lib/THC/THCTensorMode.cuh

```diff
@@ -0,0 +1,282 @@
+// buffers of Power2Size, which must be greater than the number of elements. Additionally,
+// there is one block for every slice to calculate the mode for, and in each block there
+// is one thread for every two elements.
+//
+// Both sorted and positions are assumed to be contiguous Tensors with the mode dimension
+// as the innermost dim, such that we can get the particular slice for a Tensor via its
+// linear block dimension * the slice size.
+template <typename T, unsigned int Power2Size>
+__global__ void computeMode(
+    T *input,
+    TensorInfo<T, unsigned int> values,
+    TensorInfo<long, unsigned int> indices,
+    long sliceSize)
+{
+  int tidx = threadIdx.x;
+  int stidx = blockDim.x + threadIdx.x; // Second Index this thread responsible for
+
+  // First, we need to calculate the offset into the sorted Tensor that represents
+  // the start of the slice for this block to calculate the mode for. This offset
+  // is a combination of the gridIndices, and the number of elements in the slice.
+  unsigned int blockId = getLinearBlockId<unsigned int>();
+  unsigned int linearOffset = blockId * sliceSize;
+
+  // The smem buffer is used to store the elements from the slice
+  __shared__ T    smem[Power2Size];
+
+  // Each thread loads up to two elements from the Tensor into shared memory
+  if (tidx < sliceSize) {
+    smem[tidx] = input[linearOffset + tidx];
+  }
+  if (stidx < sliceSize) {
+    smem[stidx] = input[linearOffset + stidx];
+  }
+
+  // The fmem buffer is used in multiple components of the kernel. First, it stores
+  // whether bmem[i] = i < sliceSize to mark the valid components in the smem buffer
+  // for sorting
+  __shared__ bool bmem[Power2Size];
+
+  bmem[tidx] = tidx < sliceSize;
+  bmem[stidx] = stidx < sliceSize;
+  __syncthreads(); // barrier for smem, bmem initialization
+
+  // First, sort the input slice in ascending order. smem contains the input
+  // elements, and bmem marks the valid indices
+  bitonicSortKeys<LTComp<T>, T, unsigned int, Power2Size>(smem, bmem, LTComp<T>());
+  __syncthreads(); // make no assumptions that the sort syncs at end
+
+  // The next step of our algorithm is performing a block-wide comparison of
+  // neighboring elements. In particular, given an sorted input slice A, we
+  // produce an output slice B, such that B[i] = 1 if A[i-i] != A[i], otherwise 0.
+  //
+  // Given the input A = [0, 0, 1, 1, 2, 2, 2, 4, 5, 6, 6, 7, 8]
+  //                 B = [1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1]
+  //
+  // We re-use the bmem buffer for this computation. In particular, we can think of
+  // B[i] = true indicating the start of a sequence of equal values in the sorted
+  // list.
+
+  if (tidx == 0) {
+    bmem[tidx] = true;  // for setting element 0
+  }
+
+  // Compares elements (0, 1), (2, 3), ... and sets 1, 3, ...
+  bmem[tidx * 2 + 1] = THCNumerics<T>::ne(smem[tidx * 2], smem[tidx * 2 + 1]); // (0, 1), (1, 2), etc.
```

Comment:
Ah nevermind about the access violation, sorry.

Yeah, saving on the syncthreads. While the compiler may not attempt to store and then re-load from shared memory (since it is not marked as `volatile`), I'd be deliberate about having the value in a register:

```
bool flag = THCNumerics<T>:: ...
bmem[tidx * 2 + 1] = flag;
cmem[tidx * 2 + 1] = !flag;
````